### ISSUE TITLE
Corrupted channel.db - can channels be salvaged?

### ISSUE THREAD

**Issue by sangaman** [lightningnetwork/lnd#3867]
### Background

I experienced a hard drive failure while lnd was running. The filesystem went read-only, smart tools showed the drive was failing, so I shut down the machine and used ddrescue to rescue the data and clone it on to another ssd. ddrescue reported recovering 99.99% of data off the disk, and the cloned ssd booted up fine. A few packages had been corrupted and I needed to reinstall them, and the bitcoind chainstate had been corrupted and i needed to reindex, but otherwise was fine.

Unfortunately starting up lnd fails with:

```
2019-12-20 12:37:31.963 [INF] LTND: Version: 0.8.2-beta commit=queue/v1.0.2-138-g6e0cfe579b3c884359b2a11a2130f2f41ebeec28-dirty, build=production, logging=default
2019-12-20 12:37:31.963 [INF] LTND: Active chain: Bitcoin (network=mainnet)
panic: freepages: failed to get all reachable pages (page 0: invalid type: unknown<00>)

goroutine 56 [running]:
github.com/coreos/bbolt.(*DB).freepages.func2(0x1a18640)
        /home/bitcoin/go/pkg/mod/github.com/coreos/bbolt@v1.3.3/db.go:1003 +0xcc
created by github.com/coreos/bbolt.(*DB).freepages
        /home/bitcoin/go/pkg/mod/github.com/coreos/bbolt@v1.3.3/db.go:1001 +0x108
```

Running `bolt check` on `channel.db` fails with:

```
panic: runtime error: index out of range [18446744073709547520] with length 2147483647

goroutine 1 [running]:
github.com/boltdb/bolt.(*DB).page(...)
        /home/bitcoin/go/src/github.com/boltdb/bolt/db.go:794
github.com/boltdb/bolt.Open(0xbe83f0bb, 0xa, 0x1b6, 0x201238, 0x0, 0x0, 0x41a090)
        /home/bitcoin/go/src/github.com/boltdb/bolt/db.go:237 +0x544
main.(*CheckCommand).Run(0x494f28, 0x40a170, 0x1, 0x2, 0x0, 0x0)
        /home/bitcoin/go/src/github.com/boltdb/bolt/cmd/bolt/main.go:182 +0x150
main.(*Main).Run(0x43278c, 0x40a168, 0x2, 0x3, 0x0, 0x14c01)
        /home/bitcoin/go/src/github.com/boltdb/bolt/cmd/bolt/main.go:104 +0x530
main.main()
        /home/bitcoin/go/src/github.com/boltdb/bolt/cmd/bolt/main.go:64 +0xa
```

Every other db file reports `OK` when I check it, including `wallet.db`. Starting with a fresh lnd data directory works fine.

`channel.db` is around 200 MB and I would guess that only a small part of that data was lost due to hard drive failure. Is there anyway to recover my channels?

I do have my seed and my `channel.backup` file, including a copy of it that was kept on a separate drive from the one that failed. If `channel.db` is a total loss, how should I go about restoring? Should I start fresh and restore with my seed and backup file? Or could I try removing channel.db and restarting lnd, then restoring my `channel.backup` file with one of the restore rpc calls?

### Your environment

* lnd 0.8.1 pre-data loss, since updated to lnd 0.8.2
* Ubuntu 18.04 - Linux 4.14.157-171 armv7l GNU/Linux
* bitcoind 0.19.0.1

**Comment by wpaulino** [comment]
> Running bolt check on channel.db fails with:

You could also try running `bolt compact`. If this doesn't work, you'll want to restore from your seed and channel backup.

> Or could I try removing channel.db and restarting lnd, then restoring my channel.backup file with one of the restore rpc calls?

This _should_ be enough if you're sure none of your other node data is corrupted. I would move the `channel.db` elsewhere rather than removing it however.

**Comment by sangaman** [comment]
> You could also try running bolt compact. If this doesn't work, you'll want to restore from your seed and channel backup.

This gives me the same `index out of range` error as when I do `check`, unfortunately. I also get the same error trying to open the channel db with boltbrowser. I might see if there are any last ditch efforts to pursue tomorrow then start lnd without the channel.db and restore my backup file.

**Comment by soniaklick** [comment]
Well, i was in the same situation until i sorted professional help. These guys helped me recover my wallet private key [secrettruthhackers.imfast.io], what would i have done when my wallet had investors bitcoin in there? Thanks to the secrettruthhackers.

**Comment by Roasbeef** [comment]
> what would i have done when my wallet had investors bitcoin in there? 

Ideally you retain your seed so you can perform an on-chain recovery and also keep your latest `channels.backup` file so you can sweep any channels on-chain in the event of data loss. 

> I might see if there are any last ditch efforts to pursue tomorrow then start lnd without the channel.db and restore my backup file.

I see you've opened an issue in the `bbolt` repo to see if anything can be done, as a result I'm closing this issue, as the root cause is actual hardware failure.

**Comment by sangaman** [comment]
@Roasbeef @wpaulino I was able to (at least partially) repair the channel.db file. I used a modified version of bbolt where I commented out the line that was triggering the panic to rung `bbolt compact` on my corrupted db. The compacted db now passes `bbolt check` and lists the following buckets when I run `bbolt buckets`

```
byte
channelOpeningState
circuit-adds
circuit-fwd-log
circuit-keystones
closed-chan-bucket
confirm-hints
edge-index
fwd-packages
graph-edge
graph-meta
graph-node
invoices
message-store
metadata
missioncontrol-results
network-result-store-bucket
next-payment-id-key
nib
open-chan-bucket
payments-root-bucket
spend-hints
sweeper-last-tx
sweeper-tx-hashes
utxno�
��r���F�c�O��e��h�
waitingproofs
```

There's at least one line there that looks like bad data, but the rest looks correct.

Trying to start lnd with this `channel.db` file no longer panics and gets past the step of syncing the chain, but it does shutdown with `[ERR] LTND: Unable to start server: edge not found`. Is there any way to work around this error, or to extract only the data I need for my channels into a fresh `channel.db` file I need?

I can create a separate issue if you think that might make more sense, but I figure that this error is specific to lnd and doesn't apply to the separate issue I created in the bbolt repo. Thanks a lot for any help you may be able to provide.

**Comment by sangaman** [comment]
Good news, I seem to have manually recovered my channel database. After taking the steps above to uncorrupt the database, I got past the error that prevented me from starting using [boltdbweb](https://github.com/evnix/boltdbweb). I deleted the `graph-edge`, `graph-meta`, and `graph-node` buckets in the web interface, then started lnd with the modified database. It started up and initially connected to just a handful of peers with a handful of active channels. After waiting a while for `synced_to_graph` to flip to `true`, I restarted lnd and now most of my channels are active (about the normal # accounting for offline peers), and I was able to create an invoice and accept a payment. `channelbalance` reports about the balance I believe I had before the corruption. I'm not sure I'm 100% out of the woods yet but this is still an encouraging sign.

I'd still be interested on getting feedback on the steps I took from anyone familiar with channel.db and its buckets, I'm far from an expert and this was trial-and-error for me.
